protocols/virtualdefund/virtualdefund_test.go

```diff
@@ -0,0 +1,200 @@
+		t.Run(msg, testUpdateAs(my))
+	}
+}
+
+func TestCrank(t *testing.T) {
+	for _, my := range allActors {
+		msg := fmt.Sprintf("testing crank as %s", my.Name)
+		t.Run(msg, testCrankAs(my))
+	}
+}
+
+func TestInvalidUpdate(t *testing.T) {
+	data := generateTestData()
+	virtualDefund := newObjective(false, data.vFixed, data.initialOutcome, big.NewInt(int64(data.paid)), 0)
+	invalidFinal := data.vFinal.Clone()
+	invalidFinal.ChannelNonce = big.NewInt(5)
+
+	signedFinal := state.NewSignedState(invalidFinal)
+
+	// Sign the final state by some other participant
+	signByOthers(alice, signedFinal)
+
+	e := protocols.ObjectiveEvent{ObjectiveId: virtualDefund.Id(), SignedStates: []state.SignedState{signedFinal}}
+	_, err := virtualDefund.Update(e)
+	if err.Error() != "event channelId out of scope of objective" {
+		t.Errorf("Expected error for channelId being out of scope, got %v", err)
+	}
+
+}
+
+func testUpdateAs(my ta.Actor) func(t *testing.T) {
+	return func(t *testing.T) {
+		data := generateTestData()
+		virtualDefund := newObjective(false, data.vFixed, data.initialOutcome, big.NewInt(int64(data.paid)), my.Role)
+		signedFinal := state.NewSignedState(data.vFinal)
+		// Sign the final state by some other participant
+		signByOthers(my, signedFinal)
+
+		e := protocols.ObjectiveEvent{ObjectiveId: virtualDefund.Id(), SignedStates: []state.SignedState{signedFinal}}
+
+		updatedObj, err := virtualDefund.Update(e)
+		updated := updatedObj.(*Objective)
+		for _, a := range allActors {
+			if a.Role != my.Role {
+				testhelpers.Assert(t, !isZero(updated.Signatures[a.Role]), "expected signature for participant %s to be non-zero", a.Name)
+			} else {
+				testhelpers.Assert(t, isZero(updated.Signatures[a.Role]), "expected signature for current participant %s to be zero", a.Name)
+			}
+		}
+		testhelpers.Ok(t, err)
+	}
+}
+
+func testCrankAs(my ta.Actor) func(t *testing.T) {
+	return func(t *testing.T) {
+		data := generateTestData()
+		virtualDefund := newObjective(true, data.vFixed, data.initialOutcome, big.NewInt(int64(data.paid)), my.Role)
+
+		updatedObj, se, waitingFor, err := virtualDefund.Crank(&my.PrivateKey)
+		testhelpers.Ok(t, err)
+		updated := updatedObj.(*Objective)
+
+		for _, a := range allActors {
+			if a.Role == my.Role {
+				testhelpers.Assert(t, !isZero(updated.Signatures[a.Role]), "expected signature for participant %s to be non-zero", a.Name)
+			} else {
+				testhelpers.Assert(t, isZero(updated.Signatures[a.Role]), "expected signature for current participant %s to be zero", a.Name)
+			}
+		}
+
+		testhelpers.Equals(t, waitingFor, WaitingForCompleteFinal)
+		signedByMe := state.NewSignedState(data.vFinal)
+		_ = signedByMe.Sign(&my.PrivateKey)
+		assertStateSentToEveryone(t, se, signedByMe, my)
+
+		// Update the signatures on the objective so the final state is fully signed
+		signedByOthers := signByOthers(my, state.NewSignedState(data.vFinal))
+		for i, sig := range signedByOthers.Signatures() {
```

Comment:
Should we just use `Update` here?

Reply:
While it'd probably be fine I figured it'd be safer for the `crank` test to not rely on the functionality of `Update`.